File: app/src/main/java/com/example/taxconnect/repository/AnalyticsRepository.java
package com.example.taxconnect.repository;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.FirebaseFirestore;

import java.util.HashMap;
import java.util.Map;

public class AnalyticsRepository {

    private static AnalyticsRepository instance;
    private final FirebaseAuth firebaseAuth;
    private final FirebaseFirestore firestore;

    private AnalyticsRepository() {
        firebaseAuth = FirebaseAuth.getInstance();
        firestore = FirebaseFirestore.getInstance();
    }

    public static synchronized AnalyticsRepository getInstance() {
        if (instance == null) {
            instance = new AnalyticsRepository();
        }
        return instance;
    }

    public void log(String event, Map<String, Object> details) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("event", event);
        payload.put("details", details != null ? details : new HashMap<>());
        if (firebaseAuth.getCurrentUser() != null) {
            payload.put("uid", firebaseAuth.getCurrentUser().getUid());
        }
        payload.put("timestamp", com.google.firebase.firestore.FieldValue.serverTimestamp());
        firestore.collection("analytics").add(payload);
    }
}

File: tax/app/src/main/java/com/example/taxconnect/data/repositories/AnalyticsRepository.kt
package com.example.taxconnect.data.repositories

import android.content.Context
import android.os.Bundle
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore

class AnalyticsRepository private constructor(context: Context?) {
    private val firebaseAuth: FirebaseAuth = FirebaseAuth.getInstance()
    private val firestore: FirebaseFirestore = FirebaseFirestore.getInstance()
    private var firebaseAnalytics: FirebaseAnalytics? = null

    init {
        if (context != null) {
            firebaseAnalytics = FirebaseAnalytics.getInstance(context.applicationContext)
        }
    }

    fun log(event: String, details: Map<String, Any>?) {
        // Log to Firestore
        val payload = HashMap<String, Any>()
        payload["event"] = event
        payload["details"] = details ?: HashMap<String, Any>()
        firebaseAuth.currentUser?.let { user ->
            payload["uid"] = user.uid
        }
        payload["timestamp"] = FieldValue.serverTimestamp()
        firestore.collection("analytics").add(payload)

        // Log to Firebase Analytics
        firebaseAnalytics?.let { analytics ->
            val bundle = Bundle()
            details?.forEach { (key, value) ->
                when (value) {
                    is String -> bundle.putString(key, value)
                    is Int -> bundle.putInt(key, value)
                    is Long -> bundle.putLong(key, value)
                    is Double -> bundle.putDouble(key, value)
                    is Boolean -> bundle.putBoolean(key, value)
                }
            }
            analytics.logEvent(event, bundle)
        }
    }

    companion object {
        @Volatile
        private var instance: AnalyticsRepository? = null

        @JvmStatic
        fun getInstance(context: Context): AnalyticsRepository {
            return instance ?: synchronized(this) {
                instance ?: AnalyticsRepository(context).also { instance = it }
            }
        }

        @JvmStatic
        fun getInstance(): AnalyticsRepository? {
            return instance
        }
    }
}
